### ISSUE TITLE
Removing terminated pod's Cgroup

### ISSUE THREAD

**Issue by dubstack** [kubernetes/kubernetes#29927]
For reference: For work in #27204 which is introducing pod level cgroups into Kubernetes. As a part of which we create cgroups for each pod and each of the pod's containers are brought under its pod cgroup. 

We need to cleanup the pod's cgroup once the pod is terminated. 
But we need to be careful about how we delete the cgroup. 

Some Points to consider: 
- Once the pod is terminated we can reduce it's cgroup's CpuShares to ensure that it doesn't take up significant CPU resources.
- We cannot  set memory limit to `0`  as there might be anonymous pages (tmpfs) that cannot be swapped out.
- Hence for memory, we need to ensure that all the pages are reclaimed before deleting the pod's cgroups, otherwise they would be simply moved to its parent cgroup which is highly undesirable and causes system memory pressure.
- We should delete the pod's cgroup only after the pod's volumes have been cleaned up because of the same reason as above.
- We need to ensure that no processes are attached to the cgroup. If any processes are found, they must be killed in-order to be able to delete the cgroup. 

On @vishh suggestions we were able to come up with the following logic for destroying pod cgroups.
1. Once the pod is terminated we reduce it's cgroup's CpuShares to 2(the lowest possible value).
2. Once a pod's volumes have been successfully cleaned up we spawn a separate go routine to destroy all the unwanted pod's cgroups.
3. The go routine is per pod cgroup and will run until the cgroups are successfully deleted.
4. As part of the goroutine we would attempt the following steps.
   - SIGKILL all the processes attached to cgroup. We repeatedly attempt killing the processes until success. If unsuccessful for 10 iterations we start logging failure messages and generate events.
   - Use cgroup's memory.force_empty interface to force the kernel to reclaim all pages.
5. We repeatedly reduce the cgroups memory usage until it falls below a small threshold(0.1 Mb seems to be a reasonable value).
6. In each iteration, if the current memory usage is x bytes we first limit the cgroup memory usage by writing x to memory.limit_in_bytes, then we write 0 to memory.force_empty to further reduce the memory usage, and we continually do this until the memory usage falls below the threshold.

Thanks to @vishh for the suggestions. This should provide a really safe method of deleting cgroups.

To implement this we would first need to add support for force_empty in libcontainer as it currently doesn't support force empty. 

Once we finalize the logic here I would be adding this note to the [proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/pod-resource-management.md).
@derekwaynecarr @Random-Liu @vishh PTAL.
Please cc anyone else who can provide feedback.

cc @kubernetes/sig-node

**Comment by vishh** [comment]
I took the liberty of editing the proposal directly. LGTM. Thanks for the write up!

**Comment by dubstack** [comment]
@vishh Thanks!

**Comment by dchen1107** [comment]
@dubstack I am not unassigning this from you. Somehow github go crazy, and remove you. Now I cannot even re-assign this back to you.

But your proposal lgtm

**Comment by yujuhong** [comment]
@dubstack needs to be in "kubernetes-collaborators" to be assigned issues/PRs.

**Comment by dubstack** [comment]
@dchen1107 @yujuhong I believe I was removed from the kubernetes org on the last day of my internship.
But I am working on this and would send out a PR soon.

**Comment by dubstack** [comment]
re above, @vishh added me to the Kubernetes org, but strangely I am still unable to assign this issue to myself.

**Comment by dubstack** [comment]
@vishh I was concerned about the following part of our logic:
> We repeatedly reduce the cgroups memory usage until it falls below a small threshold(0.1 Mb seems to be a reasonable value).
Once we write 0 to the force_empty file :
` almost all pages tracked by this memory cgroup will be unmapped and freed.
Some pages cannot be freed because they are locked or in-use. Such pages are moved to parent(if use_hierarchy==1) or root (if use_hierarchy==0) and this cgroup will be empty.`

So once force_empty file is successfully updated it guarantees that the cgroup would be empty. So we can just once write 0 to force_empty, and then go ahead and delete the cgroup.

**Comment by yujuhong** [comment]
@dubstack I think you can request to join kubernetes-collaborator so that you can be assigned issues.

**Comment by dubstack** [comment]
@yujuhong oh Thanks. Done!

**Comment by vishh** [comment]
@dubstack good catch!

Another option that I remember mentioning is that of decreasing
limit_in_bytes in small chunks until usage is `0` or equal to
`kmem.usage_in_bytes`.

**Comment by dubstack** [comment]
@vishh Ideally before deleting a cgroup we would want to ensure that the usage_in_bytes of the cgroup becomes 0 or falls below a certain threshold. But its not always possible to remove all charges to the cgroup and as @adityakali told me offline moving `some charges` to the parent cgroup is unavoidable. So I am not sure if progressively decreasing the limit_in_bytes in small chunks would eventually guarantee the above in a reasonable amount of time. We risk running a large number of background go-routines with no guarantees on end times. WDYT?

**Comment by hotpxl** [comment]
@dubstack I ran into this post when I was working on some CGroup issue. What do you mean by
> But its not always possible to remove all charges to the cgroup

I ran into similar issues, where `tasks` file is empty but `usage_in_bytes` is non-zero.
Could you point me to the root cause? Thanks!

**Comment by michelgokan** [comment]
This is an important issue as it makes cadvisor report tons of unnecessary information after a while (in my case cadvisor report reached 2.6MB!. It also makes housekeeping very costly.

**Comment by YueHonghui** [comment]
Any progress about this problem ? We are facing the same problem about cadvisor with @michelgokan .

**Comment by michelgokan** [comment]
I'm not 100% sure but last time I checked with the latest version of
Kubernetes (around one and half months ago) I didn't see this problem
anymore. Can you please update your Kubernetes and check if this problem
still exists or not?

**Comment by Davidrjx** [comment]
any solutins? could anyone make reply?

**Comment by YueHonghui** [comment]
@michelgokan @Davidrjx The version of kubernetes we used is based on v1.13.2, and still have the problem described here [cadvisor/issues/1774](https://github.com/google/cadvisor/issues/1774). I inspected code base of v1.13.2 and latest version on master, but didn't see any cleanups related to `memory.force_empty` or progressively lower `memory.limit_in_bytes`. 

@dubstack I don't see progress of cleanups like your proposal in libcontainer yet. Have this proposal been given up? We are facing the problem and want to have a try to fix it, any suggestions?
